### PR TITLE
[Draft] Solution submission

### PR DESCRIPTION
**Pull request by josojo**
PR only for early revision...
Also, I will split this PR into smaller chunks tomorrow.

This Pr provides the first `solutionSubmission` functionality.

It does the following for sellOrders only:
- updates all touched orders
- updates all balances
- restores old states from previous solutions
- stores updates for recreation of old solutions

it does not calculate any utility, does not check prices conformities or norms

### CODE REVIEW

File: contracts/StablecoinConverter.sol [hunk 1]
```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                "sellTokenPriceIndex of order is incorrect"
+            );
+            require(
+                tokenIdForPrice[findPriceIndex(order.buyToken, tokenIdForPrice)] == order.buyToken,
+                "BuyTokenPriceId of order is incorrect"
+            );
+            require(order.validFrom <= batchIndex, "Order is not yet valid");
+            require(order.validUntil >= batchIndex, "Order is no longer valid");
+            require(
+                prices[order.sellToken].mul(order.buyAmount) >= prices[order.buyToken].mul(order.sellAmount),
```
Comment (josojo) [inline contracts/StablecoinConverter.sol:158]: check limit prices with buyVolume and sellVolume

File: contracts/StablecoinConverter.sol [hunk 2]
```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
...
+        );
+        // checkTradersUtility();
+        // checkPriceConformity(prices);
+        // checkNorm(prices);
+        int[] memory tokenConservation = new int[](prices.length);
+        undoPreviousSolution(batchIndex);
+        uint len = owner.length;
+        for(uint i = 0; i < len; i++){
+            Order memory order = orders[owner[i]][orderId[i]];
+            require(
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol:147]: How could this ever be false?

File: contracts/StablecoinConverter.sol [hunk 3]
```diff
@@ -116,4 +119,147 @@ contract StablecoinConverter is EpochTokenLocker {
...
+                prices[order.sellToken].mul(order.buyAmount) >= prices[order.buyToken].mul(order.sellAmount),
+                "limit price not met"
+            );
+            // Assume for now that we always have sellOrders if (!order.isSellOrder) {
+            uint sellVolume = volume[i];
+            uint buyVolume = volume[i].mul(prices[findPriceIndex(order.buyToken, tokenIdForPrice)]) /
+                prices[findPriceIndex(order.sellToken, tokenIdForPrice)];
+            require(order.sellAmount >= sellVolume, "sellVolume bigger than specified in order");
+            orders[owner[i]][orderId[i]].buyAmount = uint128(
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol:166]: Not sure where this mix of integer width comes from but should the volumes than also be provided in u128?

**Review by fleupold — COMMENTED** [review]
Looking forward to the break-up. This is a lot to digest/review.

**Comment by josojo** [comment]
no longer relevant

**Pull request closed without merging**